File: 程序员面试金典第6版/面试题 10.10. Rank from Stream LCCI.java
class StreamRank {
    List<Integer> list;

    public StreamRank() {
        list = new ArrayList<Integer>();
    }
    
    public void track(int x) {
        int idx = Collections.binarySearch(list, x);
        if(idx < 0) {     // 返回负的应插入索引+1 -(idx+1) 相等插入后是第几个元素的负值
            idx = -idx-1; // 还原成正的索引位置
        }
        list.add(idx, x);
    }
    
    public int getRankOfNumber(int x) {
        int idx = Collections.binarySearch(list, x);
        if(idx < 0) {
            idx = -idx-1;
        }
        while(idx < list.size() && list.get(idx) <= x) { // 用于处理重复相等元素存在 不会调用太多次
            idx++;
        }
        return idx;
    }
}

/**
 * Your StreamRank object will be instantiated and called as such:
 * StreamRank obj = new StreamRank();
 * obj.track(x);
 * int param_2 = obj.getRankOfNumber(x);
 */

// 自定义BST

 class StreamRank {
    private class BST {
        private class BSTreeNode {
            BSTreeNode left;
            BSTreeNode right;
            int val, leftSize=1;

            public BSTreeNode(int val) {
                this.val = val;
            }

            public void insert(int val) {
                if(this.val >= val) {
                    this.leftSize++;
                    if(this.left == null) {
                        this.left = new BSTreeNode(val);
                    } else {
                        this.left.insert(val);
                    }
                } else {
                    if(this.right == null) {
                        this.right = new BSTreeNode(val);
                    } else {
                        this.right.insert(val);
                    }
                }
            }

            public int find(int val) {
                if(this.val > val) {
                    if(this.left == null) {
                        return 0;
                    } else {
                        return this.left.find(val);
                    }
                } else {
                    if(this.right == null) {
                        return this.leftSize;
                    } else {
                        return this.leftSize + this.right.find(val);
                    }
                }
            }
        }

        BSTreeNode root;
        public void insert(int val) {
            if(root == null) {
                root = new BSTreeNode(val);
            } else {
                root.insert(val);
            }
        }

        public int find(int val) {
            if(root == null) return 0;
            else return root.find(val);
        }
    } 

    BST bst;
    public StreamRank() {
        bst = new BST();
    }
    
    public void track(int x) {
        bst.insert(x);
    }
    
    public int getRankOfNumber(int x) {
        return bst.find(x);
    }
}

/**
 * Your StreamRank object will be instantiated and called as such:
 * StreamRank obj = new StreamRank();
 * obj.track(x);
 * int param_2 = obj.getRankOfNumber(x);
 */

File: 程序员面试金典第6版/面试题 10.10. Rank from Stream LCCI.cpp
// 树状数组 模板题 求和变计数
// 详见https://www.cnblogs.com/xenny/p/9739600.html
// 2^k = i&(i^(i-1)) k表示i二进制中从低到高连续零的长度
// 2^k = i&(-i) lowbit求2^k专用称呼
// x&(-x) x为零时结果为零；x为奇数时结果为1；x为偶数时，结果为x中2的最大次方的因子

// class StreamRank {
// private:
//     const static int SIZE = 5e4+10;
//     int tree[SIZE];
//     // int idx

//     int lowbit(int x) {
//         return x&(-x);
//     }

// public:
//     StreamRank() {
//         memset(tree, 0, sizeof(tree)); // fill(tree, tree+SIZE, 0);
//         // idx = 0;
//     }
    
//     void track(int x) {
//         for(int i=x+1; i<=SIZE; i+=lowbit(i)) {
//             tree[i] += 1;
//         }
//     }
    
//     int getRankOfNumber(int x) {
//         int res = 0;
//         for(int i=x+1; i>0; i-=lowbit(i)) {
//             res += tree[i];
//         }
//         return res;
//     }
// };

class StreamRank {
private:
    const static int SIZE = 5e4+10;
    int tree[SIZE];

    int lowbit(int x) {
        return x&(-x);
    }
public:
    StreamRank() {
        memset(tree, 0, sizeof(tree));
    }
    
    void track(int x) {
        for(int i=x+1; i<SIZE; i += lowbit(i)) { // +1 x可以等0 但树状数组0位置不存元素 所以全体右移1个索引
            tree[i] += 1;
        }
    }
    
    int getRankOfNumber(int x) {
        int res = 0;
        for(int i=x+1; i>0; i -= lowbit(i)) {
            res += tree[i];
        }
        return res;
    }
};

/**
 * Your StreamRank object will be instantiated and called as such:
 * StreamRank* obj = new StreamRank();
 * obj->track(x);
 * int param_2 = obj->getRankOfNumber(x);
 */